### ISSUE TITLE
TBG Enhancement #638 (Octopuss on 2014-05-19): Conflict resolution feature reporting nonexistant problems

### ISSUE THREAD

**Issue by TanninOne** [TanninOne/modorganizer#109]
I am using MO 1.2.1.
The conflict resolution feature is telling me to reorder several mods which - as far as I can tell - have no scripts in common at all.

Move Proudspire Manor Refurbished after Cutting Room Floor
Move Traps Make Noise after Proudspire Manor Refurbished
Move Enhanced Blood Textures after Traps Make Noise
Move Complete Crafting Overhaul Remade after Enhanced Blood Textures
Move When Vampires Attack after Complete Crafting Overhaul Remade
Move The Choice is Yours after When Vampires Attack
Move Thieves Guild Requirements - No Auto Quest Start Brynjolf after The Choice is Yours
Move Dragon Soul Absorb More Glorious after Thieves Guild Requirements - No Auto Quest Start Brynjolf
Move Even Better Quest Objectives after Dragon Soul Absorb More Glorious
Move Arissa - The Wandering Rogue after Even Better Quest Objectives
Move The Paarthurnax Dilemma after Arissa - The Wandering Rogue
Move Stealth Skills Rebalanced - Sneak - Lockpicking - Pickpocket after The Paarthurnax Dilemma
Move The Choice is Yours - Dawnguard after Stealth Skills Rebalanced - Sneak - Lockpicking - Pickpocket

[http://www.nexusmods.com/skyrim/mods/14630/? Proudspire Manor Refurbished]
[http://www.nexusmods.com/skyrim/mods/47327/? Cutting Room Floor]
[http://www.nexusmods.com/skyrim/mods/31799/? Traps Make Noise]
[http://www.nexusmods.com/skyrim/mods/60/? Enhanced Blood Textures]
[http://www.nexusmods.com/skyrim/mods/49791/? Complete Crafting Overhaul Remade]
[http://www.nexusmods.com/skyrim/mods/28235/? When Vampires Attack]
[http://www.nexusmods.com/skyrim/mods/26359/? The Choice is Yours]
[http://www.nexusmods.com/skyrim/mods/14157/? Thieves Guild Requirements - No Auto Quest Start Brynjolf]
[http://www.nexusmods.com/skyrim/mods/13850/? Dragon Soul Absorb More Glorious]
[http://www.nexusmods.com/skyrim/mods/32695/? Even Better Quest Objectives]
[http://www.nexusmods.com/skyrim/mods/53754/? Arissa - The Wandering Rogue]
[http://www.nexusmods.com/skyrim/mods/18465/? The Paarthurnax Dilemma]
[http://www.nexusmods.com/skyrim/mods/28418/? Stealth Skills Rebalanced - Sneak - Lockpicking - Pickpocket]

If you want my complete mod list and load order, I can post it too.

-

**Comment by TanninOne** [comment]
Octopuss on 2014-05-19: The issue was updated with the following change(s):
* This issue's description has been changed

**Comment by TanninOne** [comment]
Octopuss on 2014-05-19: Bleh, I am sorry, the editor completely messes up formatting for some reason. I can't fix it.

**Comment by TanninOne** [comment]
Octopuss on 2014-05-19: I've just noticed most of the reported problems disappear when I disable the unofficial patches. That was totally not expected.

**Comment by TanninOne** [comment]
Tannin on 2014-05-19: The reason this happens is this:
Proudspire Manor Refurbished conflicts with Unofficial Skyrim Patch
Cutting Room Floor Traps conflicts with Unofficial Skyrim Patch

That puts all three mods into one dependency graph. MO always makes suggestions based on whole dependency graphs, even if the two mods mentioned don't conflict directly.
This is why disabling one mod can make move suggestions between two other mods disappear.

**Comment by TanninOne** [comment]
Octopuss on 2014-05-19: Ah ok. Perhaps you could slightly change the description then to explain how exactly it works. Current formulation is a bit unclear.

**Comment by TanninOne** [comment]
EsJa4LjhERVzTmnj on 2014-05-25: Had noticed something similar as Octopuss, wanted to create a ticket but found that one already existed. Adding my edited 'ticket' here. 

Doing some tests with the MO warning "Potential Mod Order Problem" (PMOP). Found something curious which I can't explain with the following in mind, and I quote you "The algorithm is very simple: If scripts from two mods conflict the scripts should be loaded in the same order as the corresponding esps."

Using MO 1.2.1 beta, LOOT 0.50 (external, kickstarted by MO). Using latest  Skyrim  1.9.32.0 and all three official Skyrim DLC's (DB, DG, HF). I've created a new profile for testing and minimized it to the following three enabled mods.

1) Unofficial Skyrim Patch - loose file version (USKP)
2) Run For Your Lives (RFYL)
3) Cutting Room Floor (CRF)

All three mods have their BSA extracted (if they had one). 
Install Order (left pane), prio from low to high (number) : USKP, RFYL and CRF.
Plugins prio sorted by LOOT: uskp ..... CRF.esp and RFYL.esp. 
I get the following PMOP warning: Move Run For Your Lives after Cutting Room Floor 

I have checked the conflics tab for USKP, RFYL and CRF. RFYL and CRF have NO conflicing scripts with each other. They both overwrite different scripts in USKP. Note: CRF.esp has USKP.esp as a master. Moving the RFYL mod in the left pane after the CRF mod removes the warning.  Disabling the USKP mod ALSO removed the warning.

TLDR: CRF and RFYL don't have conflicting scripts (with each other) but they both overwrite scripts in USKP. Getting a PMOP warning. 

Given your reply #4 I can now understand why I was getting this warning. 

Am I correct when I think that in my case a PMOP warning wasn't needed and that if an other algorithm (probably more complex) was used no warning would have been shown? I think its best to have as few warnings as possible.

In this case it doesn't matter that much but when you have more mods (248 actives esm/esps currently) it does matter. The algorithm seems to give warnings and suggestions which are not always needed. Which causes unneeded left pane manipulations. If a mods have more assets then scripts, eg textures it could mean Skyrim looks or behaves different. 

Just wondering: would it be technically possible to put the algorithm in a plugin (if its not already in there). This would enable others users to create a plugin with a different algorithm. I know I don't think its very likely anyone would be doing this but just asking from a technical point of view.

To end this post. Please do NOT interpret this as criticism. I know that the whole BSA extraction, PMOP warning issues seems to give you less then positive feelings. I really hope you continue enjoying coding MO so that we selfish users can continue enjoying MO ;-)

**Comment by TanninOne** [comment]
Tannin on 2014-05-25: The algorithm is already in a plugin (diagnosebasic) so if anyone want's to work on it: be my guest.

But you should know that this is more complicated than one might initially suspect. The intuitive approach would be to work on each conflict directly, i.e. to treat the conflict "USKP vs RFYL" independent from "USKP vs CRF". but then you might end in a situation where the suggestion go in circles.

i.e. Mod A has script Conflict with B and A conflicts with C
plugin order is C, A, B
mod order is B, C, A
There are two dependencies now:
# "A before B"
# "C before A"
the latter one is already set up correctly so so the only suggestion might be: "move A before B" or "move B after A". The latter would be perfect but how would MO know which of the two to choose without looking at all dependencies?
Following the former suggestion you end up with A, B, C which breaks dependency 2 so you get a follow up warning.
Again MO might suggest to "move C before A" or "move A after C". If it suggests the former, everything is ok but you had one warning too many. If it suggests the latter you have B, C, A which is exactly where we started.
Add more mods and dependencies to the mix and things might become very complex.

To make correct suggestions MO would probably have to determine all possible solutions to the ordering and suggest the one that requires the smallest number of changes but the number of possible solutions might grow very very large.

**Comment by TanninOne** [comment]
EsJa4LjhERVzTmnj on 2014-05-25: Tannin thanks for the response. It makes it clearer now why you have taken the current approach. You are the type of person who really thinks about a problem and who looks for 'elegant' solutions. I've done enough programming in my life to know that when things look relative easy at first glance they are probably a PITA to code. 

Its not totally unlikely that this kind of 'problem'/puzzle or something similar exists for other application/ real life situations as well and that there has been done extensive research for it and that standard algorithms exists for it. After all it seems to boil down to graph theory (or trees) and that's in use in lots of (un)expected places and lots of algorithms exist for it. Still having fond memories (NOT) of my graph/tree theory courses at the university 20 years ago ;-( I would be the last to say I've a better algorithms because like you explained, its more tricky then one would expect......

**Comment by TanninOne** [comment]
Tannin on 2014-05-25: '''wolverine2710@gmail.com wrote:'''
>Tannin thanks for the response. It makes it clearer now why you have taken
>the current approach. You are the type of person who really thinks about a
>problem and who looks for 'elegant' solutions. I've done enough programming
>in my life to know that when things look relative easy at first glance they
>are probably a PITA to code.  
> 
>Its not totally unlikely that this kind of 'problem'/puzzle or something
>similar exists for other application/ real life situations as well and that
>there has been done extensive research for it and that standard algorithms
>exists for it. After all it seems to boil down to graph theory (or trees)
>and that's in use in lots of (un)expected places and lots of algorithms
>exist for it. Still having fond memories (NOT) of my graph/tree theory
>courses at the university 20 years ago ;-( I would be the last to say I've
>a better algorithms because like you explained, its more tricky then one
>would expect...... 


Well, unfortunately my knowledge of graph theory is shaky at best, a lot of the current code is trial and error.

What currently happens is:
* a) create one "vertex" per mod, including information about its current mod order, name of one of its plugins and plugin order
* b add directed edges for dependencies (if two mods share at least one script there is a dependency, direction is based on the plugin order)
* c) split up the graph into sub-graphs that have no connecting edges
* d) do a topological sort on each subgraph
* for each subgraph:
* e) go through the mods and find the longest sequence of mods that are already in the right order
* f) insert the mods not inside the sorted subgraph at the correct location
* g) print out a readable description of that insertion

Now the problem about this is that the topological sort creates a single order for each sub graph (based on plugin order) whereas in practice there are multiple options (i.e. in the USKP, RYFL, CRF case it doesn't matter if RYFL comes before CRF or after because these two aren't connected).

I could determine an order that allows multiple vertices to have the same priority: "for each vertex: sort_order = max(sort_order of all incoming edges) + 1".

This way, USKP would have sort_order 0, RYFL and CRF would both have sort_order 1.
And then, when re-ordering the mods in step e-f I would treat the sequence as sorted correctly as long as the sort_order of the mods is monotonically increasing (doesn't have to be strictly increasing then).
This way if the plugin order is USKP, RYFL, CRF and the mod order is USKP, CFL, RYFL MO would still consider the mods sorted correctly and not suggest any moves.

I '''think''' this might solve the issue but I'm not sure this is correct.

**Comment by TanninOne** [comment]
whatrevolution on 2014-06-13: '''Tannin wrote:'''
>But you should know that this is more complicated than one might initially
>suspect. The intuitive approach would be to work on each conflict directly,
>i.e. to treat the conflict "USKP vs RFYL" independent from "USKP vs CRF".
>but then you might end in a situation where the suggestion go in circles. 

Agreed wholly. On comment 2 of ticket #503 I give a current real example of this problem.

Many of these dependency/conflict networks are unsolvable by what MO is technically capable of doing, ever. That is, it's relatively simple for MO to observe some data about mod A, B, C, and D, then note that each later one replaces assets that doesn't end in a clean fit. What MO can't do is look inside the assets to figure out which internal pieces to keep, and then keep only those bits. That's a job for something like TES5Edit or CreationKit.

So it isn't generally true that MO is complaining about nonexistant problems. It's far more likely that seeing MO behave in the described way is a signal that deep problems exist in the given mod set, and that MO isn't even telling you half of the story.

**Comment by TanninOne** [comment]
Octopuss on 2014-09-01: I think this could be fixed, because there's probably nothing meaningful to be discussed here :)

**Comment by TanninOne** [comment]
Octopuss on 2014-09-10: ...lol I meant to write closed.

**Comment by TanninOne** [comment]
Tannin on 2015-01-24: I'm changing this into an enhancement request because the behaviour isn't really unexpected. Personally I'm not going to work on this any time soon, if someone else wants to improve this, the comments should contain some useful information.

**Comment by TanninOne** [comment]
Tannin on 2015-01-24: The issue was updated with the following change(s):
* The issue type has been updated, from '''Bug report''' to '''Enhancement'''.
